### PR TITLE
[Packaging] Add Ubuntu 24.04 Noble Numbat support

### LINKED ISSUE #28872: Unable to install for Ubuntu 24.04
### Describe the bug

With Ubuntu 24.04 being released, I first checked to see if `noble` was available in https://packages.microsoft.com/repos/azure-cli/dists/. 

Since it is not available yet, I tried to use the install script instead `curl -L https://aka.ms/InstallAzureCli | bash`.

Using the defaults for everything in the script (not specifying any strange paths for install directory or bin directory), I receive an error.

### Related command

```
curl -L https://aka.ms/InstallAzureCli | bash
```

### Errors

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  1405  100  1405    0     0   1981      0 --:--:-- --:--:-- --:--:--  228k
Downloading Azure CLI install script from https://azurecliprod.blob.core.windows.net/install.py to /tmp/azure_cli_install_tmp_DELCpg.
######################################################################## 100.0%
/tmp/azure_cli_install_tmp_DELCpg: OK
Running install script.
-- Verifying Python version.
-- Python version 3.12.3 okay.
-- Verifying native dependencies.
-- Unable to verify native dependencies. dist=ubuntu, version=24.04. Continuing...

===> In what directory would you like to place the install? (leave blank to use '/home/wdixon/lib/azure-cli'):
-- Creating directory '/home/wdixon/lib/azure-cli'.
-- We will install at '/home/wdixon/lib/azure-cli'.

===> In what directory would you like to place the 'az' executable? (leave blank to use '/home/wdixon/bin'):
-- Creating directory '/home/wdixon/bin'.
-- The executable will be in '/home/wdixon/bin'.
-- Downloading virtualenv package from https://pypi.python.org/packages/source/v/virtualenv/virtualenv-16.7.11.tar.gz.
-- Downloaded virtualenv package to /tmp/tmp1d0rj52p/virtualenv-16.7.11.tar.gz.
-- Checksum of /tmp/tmp1d0rj52p/virtualenv-16.7.11.tar.gz OK.
-- Extracting '/tmp/tmp1d0rj52p/virtualenv-16.7.11.tar.gz' to '/tmp/tmp1d0rj52p'.
-- Executing: ['/usr/bin/python3', 'virtualenv.py', '--python', '/usr/bin/python3', '/home/wdixon/lib/azure-cli']
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/wdixon/lib/azure-cli/bin/python3
Also creating executable in /home/wdixon/lib/azure-cli/bin/python
ERROR: The executable /home/wdixon/lib/azure-cli/bin/python3 is not functioning
ERROR: It thinks sys.prefix is '/usr' (should be '/home/wdixon/lib/azure-cli')
ERROR: virtualenv is not compatible with this system or executable
Traceback (most recent call last):
  File "/tmp/azure_cli_install_tmp_DELCpg", line 415, in <module>
    main()
  File "/tmp/azure_cli_install_tmp_DELCpg", line 399, in main
    create_virtualenv(tmp_dir, install_dir)
  File "/tmp/azure_cli_install_tmp_DELCpg", line 144, in create_virtualenv
    exec_command(cmd, cwd=working_dir)
  File "/tmp/azure_cli_install_tmp_DELCpg", line 105, in exec_command
    subprocess.check_call(command_list, cwd=cwd, env=env)
  File "/usr/lib/python3.12/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', 'virtualenv.py', '--python', '/usr/bin/python3', '/home/wdixon/lib/azure-cli']' returned non-zero exit status 100.
```

### Issue script & Debug output

N/A

### Expected behavior

I would expect to either setup the PPA or use the Install script to successfully get a working version of the Azure CLI.

### Environment Summary

```
$ python3 --version
Python 3.12.3

$ which python3
/usr/bin/python3

$ cat /etc/os-release
PRETTY_NAME="Ubuntu 24.04 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bebound**
Ubuntu 24.04 was released on April 25, 2024
Resolve #28872

Ref: https://wiki.ubuntu.com/Releases

### CODE REVIEW

**Comment by yonzhan** [comment]
Packaging

**Comment by bebound** [comment]
Noble package is half the size of Jammy:
```
-rwxrwxrwx 1 root root 54M May  8 11:22 azure-cli_2.60.0-1~jammy_amd64.deb
-rwxrwxrwx 1 root root 27M May  8 10:59 azure-cli_2.60.0-1~noble_amd64.deb
```

`Installed-Size: 697 MB` vs `Installed-Size: 359 MB`

Some `__pycache__` folders are missing in Noble, for example, `/opt/az/lib/python3.11/site-packages/azure/keyvault/keys/__pycache__`

**Comment by bebound** [comment]
Noble uses `dpkg` 1.22.6, while Jammy uses 1.21.1.
The `dh_clean` step is slightly different while running `dpkg-buildpackage`.
`-name __pycache__` is added in the `find` command and they are removed from final package.

**This would significantly slow the `az` command from 2s to 11s for non-root user, which can't generate `__pycache__` folder during command execution.** 

```
dh_clean: warning: Compatibility levels before 10 are deprecated (level 9 in use)
 	rm -fr -- debian/azure-cli/ debian/tmp/
 	find .  \( \( \
 		\( -path .\*/.git -o -path .\*/.svn -o -path .\*/.bzr -o -path .\*/.hg -o -path .\*/CVS -o -path .\*/.pc -o -path .\*/_darcs \) -prune -o -type f -a \
 	        \( -name '#*#' -o -name '.*~' -o -name '*~' -o -name DEADJOE \
 		 -o -name '*.orig' -o -name '*.rej' -o -name '*.bak' \
 		 -o -name '.*.orig' -o -name .*.rej -o -name '.SUMS' \
 		 -o -name TAGS -o \( -path '*/.deps/*' -a -name '*.P' \) \
 		\) -exec rm -f {} + \) -o \
 		\( -type d -a \( -name autom4te.cache -o -name __pycache__ \) -prune -exec rm -rf {} + \) \)

```

File: scripts/release/debian/prepare.sh [hunk 1]
```diff
@@ -102,6 +102,9 @@ export DH_OPTIONS=-v
 %:
 ${TAB}dh \$@ --sourcedirectory $source_dir
 
+override_dh_clean:
+${TAB}dh_clean --exclude=__pycache__
```
Comment (bebound) [inline scripts/release/debian/prepare.sh:106]: To ensure the pyc files are included in package, as they are excluded by default starting with debhelper version 13.10.10.
See https://github.com/Azure/azure-cli/pull/28888#issuecomment-2099909964

**Comment by bebound** [comment]
As suggest by Niels from Debian team, we should move the byte-compile step in the post install script, which could help us to reduce the package size.

Debian Python Policy also suggests this:
> If a binary package provides any binary-independent modules (foo.py files), the corresponding byte-compiled modules (foo.pyc files) and optimized modules (foo.pyo files) must not ship in the package. Instead, they should be generated in the package’s post-install script, and removed in the package’s pre-remove script.   -- https://www.debian.org/doc/packaging-manuals/python-policy/#modules-byte-compilation

This is beyond the scope of this PR. I've simply omitted the `__pycache__` file from `dh_clean` during packaging.

File: scripts/release/debian/prepare.sh (hunk 1 above)
Comment (jiasli) [reply scripts/release/debian/prepare.sh:106]: This is a more official doc: https://manpages.ubuntu.com/manpages/noble/en/man1/dh_clean.1.html

**Review by jiasli — APPROVED** [review]

**Comment by jiasli** [comment]
Fascinating in-depth investigation!

**Pull request merged**